torch/optim/lr_scheduler.py

```diff
@@ -427,3 +427,89 @@ def state_dict(self):
     def load_state_dict(self, state_dict):
         self.__dict__.update(state_dict)
         self._init_is_better(mode=self.mode, threshold=self.threshold, threshold_mode=self.threshold_mode)
+
+
+class SGDR(_LRScheduler):
+    r"""Set the learning rate of each parameter group using a cosine annealing
+    schedule, where :math:`\eta_{max}` is set to the initial lr, :math:`T_{cur}`
+    is the number of epochs since the last restart and :math:`T_{i}` is the number
+    of epochs between two warm restarts in SGDR:
+
+    .. math::
+
+        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
+        \cos(\frac{T_{cur}}{T_{i}}\pi))
+
+    When :math:`\T_{cur}=T_{i}`, set :math:`\eta_t = \eta_{min}`.
+    When :math:`\T_{cur}=0`(after restart), set :math:`\eta_t=\eta_{max}`.
+
+    It has been proposed in
+    `SGDR: Stochastic Gradient Descent with Warm Restarts`_.
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        T_0 (int): Number of iterations for the first restart.
+        T_mult (float): A factor increases :math:`\T_{i}` after a restart. Default: 1.
+        eta_min (float): Minimum learning rate. Default: 0.
+        last_epoch (int): The index of last epoch. Default: -1.
+
+    .. _SGDR\: Stochastic Gradient Descent with Warm Restarts:
+        https://arxiv.org/abs/1608.03983
+    """
+
+    def __init__(self, optimizer, T_0, T_mult=1, eta_min=0, last_epoch=-1):
+        if T_0 < 0:
+            raise ValueError("Expected positive T_0, but got {}".format(T_0))
+        if T_mult < 0:
+            raise ValueError("Expected positive T_mul, but got {}".format(T_mul))
+        self.T_0 = int(T_0)
+        self.T_i = int(T_0)
+        self.T_mult = T_mult
+        self.eta_min = eta_min
+        super(SGDR, self).__init__(optimizer, last_epoch)
+        for i, group in enumerate(optimizer.param_groups):
+            if 'initial_lr' not in group:
+                raise KeyError("param 'initial_lr' is not specified "
+                               "in param_groups[{}] when resuming an optimizer".format(i))
```

Comment:
is this `resuming` or `assuming`?

Reply:
I copied the line 23 from the base class `_LRScheduler(object)`. In my opinion, it is `resuming`. But I think these codes are redundant, the `optimizer` will be checked by calling `super(SGDR, self).__init__(optimizer, last_epoch)`